From e122864eec6b074cd318a6b287af6b77f73a1ce2 Mon Sep 17 00:00:00 2001
From: Paul Reioux <reioux@gmail.com>
Date: Sun, 7 Sep 2014 23:55:10 -0500
Subject: [PATCH] intelli-thermal v2: emulate existing behavior

Signed-off-by: Paul Reioux <reioux@gmail.com>
---
 drivers/thermal/intelli/msm_thermal_v2.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/thermal/intelli/msm_thermal_v2.c b/drivers/thermal/intelli/msm_thermal_v2.c
index af999cf..9644a96 100644
--- a/drivers/thermal/intelli/msm_thermal_v2.c
+++ b/drivers/thermal/intelli/msm_thermal_v2.c
@@ -1942,6 +1942,12 @@ int __devinit msm_thermal_init(struct msm_thermal_data *pdata)
 	if (num_possible_cpus() > 1)
 		register_cpu_notifier(&msm_thermal_cpu_notifier);
 
+	/* emulate default behavior */
+	disable_msm_thermal();
+	hotplug_init();
+	freq_mitigation_init();
+	enabled = 0;
+
 	return ret;
 }
 
