From 31654d007cc424197ad9cafbc91cb904805810dd Mon Sep 17 00:00:00 2001
From: friedrich420 <kelpidorou@yahoo.com>
Date: Sun, 7 Feb 2016 14:44:27 +0200
Subject: [PATCH] fs/dcache.c: Increase the too small buffer for dname [Devin
 Kim]

---
 fs/dcache.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/dcache.c b/fs/dcache.c
index 105db7d..fd82e18 100644
--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -2726,7 +2726,7 @@ char *dynamic_dname(struct dentry *dentry, char *buffer, int buflen,
 			const char *fmt, ...)
 {
 	va_list args;
-	char temp[64];
+	char temp[256];
 	int sz;
 
 	va_start(args, fmt);
