From 43d4d0af7711d16df6126fc575e421e791211712 Mon Sep 17 00:00:00 2001
From: Hybridmax <hybridmax95@gmail.com>
Date: Sun, 8 Mar 2015 18:44:15 +0100
Subject: [PATCH] cfg80211: Fix memory leak

---
 net/wireless/reg.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/net/wireless/reg.c b/net/wireless/reg.c
index f3090b7..5049344 100755
--- a/net/wireless/reg.c
+++ b/net/wireless/reg.c
@@ -2236,10 +2236,15 @@ static int __set_regdom(const struct ieee80211_regdomain *rd)
 		 * However if a driver requested this specific regulatory
 		 * domain we keep it for its private use
 		 */
-		if (last_request->initiator == NL80211_REGDOM_SET_BY_DRIVER)
+		if (last_request->initiator == NL80211_REGDOM_SET_BY_DRIVER) {
+			const struct ieee80211_regdomain *tmp;
+
+			tmp = request_wiphy->regd;
 			request_wiphy->regd = rd;
-		else
+			kfree(tmp);
+		} else {
 			kfree(rd);
+		}
 
 		rd = NULL;
 
