From 4eff5a02366057f0576c14a2263d2a04d0739ffa Mon Sep 17 00:00:00 2001
From: shminer <332574643@qq.com>
Date: Wed, 15 Apr 2015 17:58:10 +0800
Subject: [PATCH] intelli_thermal:show sensor id for thermal tweak

---
 drivers/thermal/intelli/msm_thermal.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/thermal/intelli/msm_thermal.c b/drivers/thermal/intelli/msm_thermal.c
index fe55d54..536e1a4 100644
--- a/drivers/thermal/intelli/msm_thermal.c
+++ b/drivers/thermal/intelli/msm_thermal.c
@@ -275,6 +275,8 @@ struct msm_thermal_stat_data {
 static struct msm_thermal_stat_data msm_thermal_stats;
 
 /* module parameters */
+module_param_named(sensor_id, msm_thermal_info.sensor_id,
+			int, 0664);
 module_param_named(poll_ms, msm_thermal_info.poll_ms, uint, 0664);
 module_param_named(limit_temp_degC, msm_thermal_info.limit_temp_degC,
 			int, 0664);
