From 5b941aa8729410eec94e7426481f9cdbc24064e1 Mon Sep 17 00:00:00 2001
From: Alucard24 <dmbaoh2@gmail.com>
Date: Sat, 22 Feb 2014 12:53:56 +0100
Subject: [PATCH] Updated fiops io-schedule. Thanks to Yuri!

---
 block/fiops-iosched.c | 15 ---------------
 1 file changed, 15 deletions(-)

diff --git a/block/fiops-iosched.c b/block/fiops-iosched.c
index 942ed53..671b1d3 100644
--- a/block/fiops-iosched.c
+++ b/block/fiops-iosched.c
@@ -408,21 +408,6 @@ static struct fiops_ioc *fiops_select_ioc(struct fiops_data *fiopsd)
 		return NULL;
 	}
 
-	/* Let sync request preempt async queue */
-	if (!rq_is_sync(rq) && service_tree->count > 1) {
-		struct rb_node *tmp = rb_next(&ioc->rb_node);
-		struct fiops_ioc *sync_ioc = NULL;
-		while (tmp) {
-			sync_ioc = rb_entry(tmp, struct fiops_ioc, rb_node);
-			rq = rq_entry_fifo(sync_ioc->fifo.next);
-			if (rq_is_sync(rq))
-				break;
-			tmp = rb_next(&sync_ioc->rb_node);
-		}
-		if (sync_ioc)
-			ioc = sync_ioc;
-	}
-
 	return ioc;
 }
 
