From 7974866b2da9266cd00c0ac1c1dbefec2cc5f06f Mon Sep 17 00:00:00 2001
From: andip71 <andreasp@gmx.de>
Date: Wed, 21 Oct 2015 07:16:16 +0200
Subject: [PATCH] random: increase entropy levels only very slightly (both read
 and write)

---
 drivers/char/random.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/char/random.c b/drivers/char/random.c
index 86a6615..0d416c8 100644
--- a/drivers/char/random.c
+++ b/drivers/char/random.c
@@ -271,9 +271,9 @@
  * /dev/random.  Should be enough to do a significant reseed.
  */
 #ifdef CONFIG_CRYPTO_FIPS
-static int random_read_wakeup_thresh = 320;
+static int random_read_wakeup_thresh = 288;
 #else
-static int random_read_wakeup_thresh = 128;
+static int random_read_wakeup_thresh = 92;
 #endif
 /*
  * If the entropy count falls under this number of bits, then we
@@ -281,9 +281,9 @@ static int random_read_wakeup_thresh = 128;
  * access to /dev/random.
  */
 #ifdef CONFIG_CRYPTO_FIPS
-static int random_write_wakeup_thresh = 512;
+static int random_write_wakeup_thresh = 352;
 #else
-static int random_write_wakeup_thresh = 256;
+static int random_write_wakeup_thresh = 160;
 #endif
 
 /*
