From 42e68a8ac16da06adb8e1c5db58cadfa407a1c13 Mon Sep 17 00:00:00 2001
From: Dorimanx <yuri@bynet.co.il>
Date: Mon, 17 Aug 2015 11:38:35 +0300
Subject: [PATCH] LMK: Enabled Adaptive LMK thanks to @neobuddy89 and tuned OOM
 levels.

---
 drivers/staging/android/lowmemorykiller.c | 22 +++++++++++-----------
 1 file changed, 11 insertions(+), 11 deletions(-)

diff --git a/drivers/staging/android/lowmemorykiller.c b/drivers/staging/android/lowmemorykiller.c
index 3d8d317..2c9aa46 100644
--- a/drivers/staging/android/lowmemorykiller.c
+++ b/drivers/staging/android/lowmemorykiller.c
@@ -72,25 +72,25 @@ static int lowmem_minfree[6] = {
 	 3 *  512,	/* Foreground App: 	6 MB	*/
 	 2 * 1024,	/* Visible App: 	8 MB	*/
 	 4 * 1024,	/* Secondary Server: 	16 MB	*/
-	16 * 1024,	/* Hidden App: 		64 MB	*/
-	28 * 1024,	/* Content Provider: 	112 MB	*/
-	32 * 1024,	/* Empty App: 		128 MB	*/
+	28 * 1024,	/* Hidden App: 		114 MB	*/
+	31 * 1024,	/* Content Provider: 	127 MB	*/
+	34 * 1024,	/* Empty App: 		139 MB	*/
 };
 static int lowmem_minfree_screen_off[6] = {
 	3 * 512,	/* 6MB */
 	2 * 1024,	/* 8MB */
 	4 * 1024,	/* 16MB */
-	16 * 1024,	/* 64MB */
-	28 * 1024,	/* 112 MB */
-	32 * 1024,	/* 128 MB */
+	28 * 1024,	/* 114MB */
+	31 * 1024,	/* 127 MB */
+	34 * 1024,	/* 139 MB */
 };
 static int lowmem_minfree_screen_on[6] = {
 	3 * 512,	/* 6MB */
 	2 * 1024,	/* 8MB */
 	4 * 1024,	/* 16MB */
-	16 * 1024,	/* 64MB */
-	28 * 1024,	/* 112 MB */
-	32 * 1024,	/* 128 MB */
+	28 * 1024,	/* 114MB */
+	31 * 1024,	/* 127 MB */
+	34 * 1024,	/* 139 MB */
 };
 static int lowmem_minfree_size = 6;
 static int lmk_fast_run = 1;
@@ -107,7 +107,7 @@ static atomic_t shift_adj = ATOMIC_INIT(0);
 static short adj_max_shift = 353;
 
 /* User knob to enable/disable adaptive lmk feature */
-static int enable_adaptive_lmk;
+static int enable_adaptive_lmk = 1;
 module_param_named(enable_adaptive_lmk, enable_adaptive_lmk, int,
 	S_IRUGO | S_IWUSR);
 
@@ -118,7 +118,7 @@ module_param_named(enable_adaptive_lmk, enable_adaptive_lmk, int,
  * 90-94. Usually this is a pseudo minfree value, higher than the
  * highest configured value in minfree array.
  */
-static int vmpressure_file_min;
+static int vmpressure_file_min = 53059;
 module_param_named(vmpressure_file_min, vmpressure_file_min, int,
 	S_IRUGO | S_IWUSR);
 
