From b0fb815570677de97cae0e5617d74ca3547a89cd Mon Sep 17 00:00:00 2001
From: robcore <robpatershuk@gmail.com>
Date: Wed, 13 Apr 2016 09:27:14 -0600
Subject: [PATCH] revert drivers/base/sync.c tweak

---
 drivers/base/sync.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/base/sync.c b/drivers/base/sync.c
index c0b9d63..dd697cc 100644
--- a/drivers/base/sync.c
+++ b/drivers/base/sync.c
@@ -79,13 +79,13 @@ static void sync_timeline_free(struct kref *kref)
 		container_of(kref, struct sync_timeline, kref);
 	unsigned long flags;
 
+	if (obj->ops->release_obj)
+		obj->ops->release_obj(obj);
+
 	spin_lock_irqsave(&sync_timeline_list_lock, flags);
 	list_del(&obj->sync_timeline_list);
 	spin_unlock_irqrestore(&sync_timeline_list_lock, flags);
 
-	if (obj->ops->release_obj)
-		obj->ops->release_obj(obj);
-
 	kfree(obj);
 }
 
